### PR TITLE
Modernize and cleanup locking

### PR DESCRIPTION
**Pull request by jimfulton**
- Move from try/finally to with

- Start phasing out the _lock_acquire/_lock_release shortcuts

- Replace simpler @locked decorators with with statements.

  I've come to prefer the with style tp the decorator style.
  I think it's clearer.

  (The decorators preceeded the Python with statement.)

  I left the decorators in cases where they were used with preconditions.

This is mostly pretty mechanical, although it got a little delicate in
places and ... tests

### CODE REVIEW

File: src/ZODB/BaseStorage.py [hunk 1]
```diff
@@ -85,13 +85,13 @@ def __init__(self, name, base=None):
 
         # Allocate locks:
         self._lock = utils.RLock()
-        self.__commit_lock = utils.Lock()
```
Comment (tseaver) [inline src/ZODB/BaseStorage.py]: Yay!

File: src/ZODB/ActivityMonitor.py [hunk 2]
```diff
@@ -41,18 +41,15 @@ def closedConnection(self, conn):
...
-        self.trim_lock.release()
+        with self.trim_lock:
+            log = self.log
+            cutoff = now - self.history_length
+            n = 0
+            loglen = len(log)
+            while n < loglen and log[n][0] < cutoff:
+                n = n + 1
+            if n:
+                del log[:n]
```
Comment (tseaver) [inline src/ZODB/ActivityMonitor.py]: Fixes a deadlock if any of that code could ever raise.

File: src/ZODB/BaseStorage.py [hunk 3]
```diff
@@ -245,8 +235,7 @@ def tpc_finish(self, transaction, f=None):
         # to send an invalidation message to all of the other
         # connections!
 
-        self._lock_acquire()
-        try:
+        with self._lock:
             if transaction is not self._transaction:
                 raise POSException.StorageTransactionError(
                     "tpc_finish called with wrong transaction")
```
Comment (tseaver) [inline src/ZODB/BaseStorage.py]: When / how does `self._commit_lock` get released in this case?

Comment (jimfulton) [inline src/ZODB/BaseStorage.py]: It gets released in the finally below.

(I may be misunderstanding your question.)

File: src/ZODB/FileStorage/fspack.py [hunk 4]
```diff
@@ -445,11 +442,10 @@ def close_files_remove():
             # pack didn't free any data.  there's no point in continuing.
             close_files_remove()
             return None
-        self._commit_lock_acquire()
+        self._commit_lock.acquire()
```
Comment (tseaver) [inline src/ZODB/FileStorage/fspack.py]: Since acquire/release all occur in the same method, wouldn't this be clearer using `with self._commit_lock:`?

File: src/ZODB/MappingStorage.py [hunk 5]
```diff
@@ -269,9 +267,9 @@ def tpc_begin(self, transaction, tid=None):
         if transaction is self._transaction:
             raise ZODB.POSException.StorageTransactionError(
                 "Duplicate tpc_begin calls for same transaction")
-        self._lock_release()
+        self._lock.release()
         self._commit_lock.acquire()
-        self._lock_acquire()
+        self._lock.acquire()
```
Comment (tseaver) [inline src/ZODB/MappingStorage.py]: Maybe the `utils.RLock` needs to grow a second context manager helper, to handle a scoped release-then-acquire.

File: src/ZODB/FileStorage/fspack.py (hunk 4 above)
Comment (jimfulton) [inline src/ZODB/FileStorage/fspack.py]: <shrug>

This packing code is very old and really ought to be updated.  But this is out of scope of this PR.

File: src/ZODB/MappingStorage.py (hunk 5 above)
Comment (jimfulton) [inline src/ZODB/MappingStorage.py]: ``utils.RLock`` is only used when debugging locks. The locks used here are regular locks normally.

This would be a little clearer if I didn't use the decorator and used 2 with's.  I decided to still with the decorators, as they're used pretty consistently in this class.

Comment (tseaver) [inline src/ZODB/MappingStorage.py]: What if we add a helper class, something like:

```python
import contextlib


@contextlib.contextmanager
def LockReleaser(real_lock):
    """Context manager:  release a lock temporarily.
    
    E.g.:

        with real_lock:
            do_some_work_with_lock()
            with LockReleaser(real_lock):
                do_work_without_lock()
            do_more_work_with_lock()
    """
    real_lock.release()
    try:
        yield None
    finally:
        real_lock.acquire()
```

Comment (jimfulton) [inline src/ZODB/MappingStorage.py]: I addressed this another way.

**Comment by jimfulton** [comment]
@tseaver are you good with this now?

File: src/ZODB/BaseStorage.py [hunk 6]
```diff
@@ -85,13 +85,13 @@ def __init__(self, name, base=None):
...
+        self._commit_lock = utils.Lock()
 
-        # Comment out the following 4 lines to debug locking:
+        # Needed by external storages that use this dumb api :(
         self._lock_acquire = self._lock.acquire
         self._lock_release = self._lock.release
-        self._commit_lock_acquire = self.__commit_lock.acquire
-        self._commit_lock_release = self.__commit_lock.release
+        self._commit_lock_acquire = self._commit_lock.acquire
+        self._commit_lock_release = self._commit_lock.release
```
Comment (tseaver) [inline src/ZODB/BaseStorage.py]: Are these micro-optimizations still worth preserving?

**Comment by tseaver** [comment]
OK to merge even without addressing my release-then-acquire concerns (it still smells like a bug magnet, though).

File: src/ZODB/BaseStorage.py (hunk 6 above)
Comment (jimfulton) [inline src/ZODB/BaseStorage.py]: They're part of the old inheritance API. :( The comment above explains this.

**Comment by jimfulton** [comment]
I thought I *did* address your concerns. Did you see the change I made?  The locked regions for the ordinary lock should be very obvious now.

File: src/ZODB/BaseStorage.py (hunk 6 above)
Comment (tseaver) [inline src/ZODB/BaseStorage.py]: Ugh, what a terrible pattern:  who could have done such a thing? (he asks innocently).

**Comment by tseaver** [comment]
The one that I commented on last me  the `_commit_lock`, which is [acquired in one method](https://github.com/zopefoundation/ZODB/pull/72/files#diff-71117ab239f0de5734a0a100e675fe14R183), held across multiple method invocations, and released later.  It is also [released-then-acquired two layers down](https://github.com/zopefoundation/ZODB/pull/72/files#diff-808cb96d0cd7582ec5350a774ad64854R637) in subclass code.

**Comment by jimfulton** [comment]
OK, wrt commit lock being held over multiple methods:

a) This is a key aspect of the two-phase commit protocol.  At least when we return from voting, we have to be assured that we can commit the modified objects, so we have to prevent modification while in the second phase.  Currently, we get the lock in tpc_begin.  This is a design flaw.  We don't really need to get the vote until tpc_vote, and of course, we should be able to get object-level locks, rather than a global lock.  This is related to https://github.com/zopefoundation/ZODB/pull/69.  So, eventually we won't get locks until tpc_vote, but even then, the lock will be acquired in tpc_vote and released in tpc_finish or tpc_abort.

b) This is out of scope of this change. :)  Although, I think using ``with``, especially when temporarily releasing the storage lock to acquire the commit lock, makes the logic a little clearer.
